cpp/src/strings/substring.cu

```diff
@@ -0,0 +1,267 @@
+ */
+enum TwoPass
+{
+    SizeOnly = 0, ///< calculate the size only
+    ExecuteOp     ///< run the string operation
+};
+
+/**
+ * @brief Function logic for the substring API.
+ * This will perform a substring operation on each string
+ * using the provided start, stop, and step parameters.
+ */
+template <TwoPass Pass=SizeOnly>
+struct substring_fn
+{
+    const cudf::column_device_view d_column;
+    int32_t start, stop, step;
+    const int32_t* d_offsets{};
+    char* d_chars{};
+
+    __device__ cudf::size_type operator()(cudf::size_type idx)
+    {
+        if( d_column.is_null(idx) )
+            return 0; // null string
+        cudf::string_view d_str = d_column.template element<cudf::string_view>(idx);
+        auto length = d_str.length();
+        if( start >= length )
+            return 0; // empty string
+        auto itr = d_str.begin() + start;
+        auto end_itr = d_str.begin() + (((stop<0) || (stop>length) ) ? length : stop);
+        cudf::size_type bytes = 0;
+        char* d_buffer = nullptr;
+        if( Pass==ExecuteOp )
+            d_buffer = d_chars + d_offsets[idx];
+        while( itr < end_itr )
+        {
+            if( Pass==SizeOnly )
+                bytes += cudf::strings::detail::bytes_in_char_utf8(*itr);
+            else
+                d_buffer += cudf::strings::detail::from_char_utf8(*itr,d_buffer);
+            itr += step;
+        }
+        return bytes;
+    }
+};
+
+} // namespace
+
+namespace cudf
+{
+namespace strings
+{
+
+//
+std::unique_ptr<cudf::column> substring( const strings_column_view strings,
+                                         size_type start, size_type stop, size_type step,
+                                         rmm::mr::device_memory_resource* mr,
+                                         cudaStream_t stream )
+{
+    size_type strings_count = strings.size();
+    if( strings_count == 0 )
+        return detail::make_empty_strings_column(mr,stream);
+    CUDF_EXPECTS( start >= 0, "Parameter start must be zero or positive integer.");
+    if( step == 0 )
+        step = 1;
+    CUDF_EXPECTS( step > 0, "Parameter step must be positive integer.");
+    if( (stop > 0) && (start > stop) )
+        CUDF_FAIL("Invalid start or stop parameter value.");
+
+    auto strings_column = column_device_view::create(strings.parent(),stream);
+    auto d_column = *strings_column;
+
+    // copy the null mask
+    rmm::device_buffer null_mask;
+    cudf::size_type null_count = d_column.null_count();
+    if( d_column.nullable() )
+        null_mask = rmm::device_buffer( d_column.null_mask(),
+                                        cudf::bitmask_allocation_size_bytes(strings_count),
+                                        stream, mr);
+    // build offsets column
+    auto offsets_transformer_itr = thrust::make_transform_iterator( thrust::make_counting_iterator<int32_t>(0),
+        substring_fn<SizeOnly>{d_column, start, stop, step} );
```

Comment:
The functor is created here without setting its internal pointer members, so they are nullptr. Where do they get set?

Reply:
They are not needed on `SizeOnly` pass just on the `ExecuteOp` pass about 13 lines below this one. So the `SizeOnly` one looks like this:
```
substring_fn<SizeOnly>{d_column, start, stop, step}
```
And the `ExecuteOp` one just below in the same function looks like this:
```
substring_fn<ExecuteOp>{d_column, start, stop, step, d_new_offsets, d_chars};
```
The `d_offsets` and `d_chars` members are not used during `SizeOnly` pass.